src/transformStateWithClones.js

```diff
@@ -7,7 +7,33 @@
  * @return {Object[]}
  */
 function transformStateWithClones(state, actions) {
-  // write code here
+  const obj = { ...state };
+  const arr = [];
+
+  for (const key of actions) {
+    if (key.type === 'addProperties') {
+      for (const add in key.extraData) {
+        obj[add] = key.extraData[add];
+      }
+      arr.push({ ...obj });
+    }
+
+    if (key.type === 'clear') {
+      for (const del in obj) {
+        delete obj[del];
+      }
+      arr.push({});
+    }
+
+    if (key.type === 'removeProperties') {
+      for (const remove of key.keysToRemove) {
+        delete obj[remove];
+      }
+      arr.push({ ...obj });
+    }
+  }
```

Comment:
You can push a copy of your current state just one time at the end of the cycle

```suggestion
    }
    
    arr.push({ ...obj });
  }
```